### ISSUE TITLE
SelectEntityActionSpace

### ISSUE THREAD

**Issue by cswinter** [entity-neural-network/incubator#19]
While working on the PPO implementation, I realized the `SelectEntityActionSpace` is extra tricky to work with because the number of choices will be different for each entity. This means we cannot easily combine all logits in a single batch into a tensor that can be passed to the torch `Categorical`. Just looping over all entities turned out to be unusable slow in the case of categorical actions (not entirely sure why but the torch `Categorical` object seems to have very high overhead). In addition, calculating the logits cannot easily be done in a batched manner either since it requires a ragged outer product operation. I think ultimately we will need to write some triton kernels for ragged ~outer product~ matmul and ragged softmax operations. Until then, it would still be good to have a workable solution that's reasonably performant. I can think of two approaches:
1. Just loop over each observation in a batch and try and make it reasonably efficient somehow. It could be that e.g. just replacing the torch `Categorical` with our own implementation and computing everything CPU-side would give us acceptable performance.
2. Pad everything out to a fixed maximum number of selectable entities.

**Comment by cswinter** [comment]
![SelectEntityAction](https://user-images.githubusercontent.com/12845088/143598833-e312dd3c-6a70-4ab1-b668-753b37a2c309.png)

**Comment by cswinter** [comment]
![SelectEntityAction v2](https://user-images.githubusercontent.com/12845088/143601909-b4aa4498-3f51-4540-960a-fee6fa9b74e5.png)

**Comment by Bam4d** [comment]
If you have many entities with different action spaces (for example in micro-RTS and some griddly envs) typically you can deal with this by flattening out all the possible actions into a single action space, and then masking off the things that are not possible for specific entities, it means you can send batches to the `Categorical` rather than iterating over each one per different entity.
